### ISSUE TITLE
Softmax doesn't support sparse tensors with the CSR layout

### ISSUE THREAD

**Issue by candyflower2005** [pytorch/pytorch#105329]
### 🚀 The feature, motivation and pitch

Hi,
both `torch.softmax` and `torch.sparse.softmax` don't support the CSR format.
Namely, when I try to apply any softmax to `torch.sparse_csr tensor`, I receive the following error:

``RuntimeError: unsupported tensor layout: SparseCsr``

Would it be possible to start supporting this format?

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer

**Comment by nikitaved** [comment]
@candyflower2005 , unfortunately, we do not have that operation, but if you have CUDA and Triton, you can convert your CSR inputs to BSR with blocksizes which are powers of 2 and greater than 16, you could use `bsr_softmax` from `torch.sparse._triton_ops`. Could you please tell more about your use case as well, especially in combination with https://github.com/pytorch/pytorch/issues/105319?

**Comment by candyflower2005** [comment]
Hi @nikitaved, thank you for the quick response. Even though the masks are quite sparse, the problem is that my sparsity pattern doesn't guarantee that there will be few non-zero blocks (that is, it's very unlikely that a block of size 16x16 will have 256 zeros), so in this situation I would prefer to keep data in the sparse CSR layout. Do you know if this feature will be available any time soon?
The second PR is for a different use case and doesn't have as high priority for me as this one.

**Comment by candyflower2005** [comment]
As negative infinity, similar to what `torch.sparse.softmax` does.

**Comment by nikitaved** [comment]
I see, then you can use `bsr_softmax` from `torch.sparse._triton_ops` after converting CSR to BSR with blocksize 1. If you use CUDA and Triton that is. The softmax kernel does not have such limitations on blocksize as mm does.

**Comment by candyflower2005** [comment]
Right, thank you.
When I try to convert a tensor from CSR to BSR:
`t = t.to_sparse_bsr(blocksize=(1,1))`
I get the following error:
`RuntimeError: to_sparse_bsr: conversion from Csr to Bsr for batched inputs is not implemented.`
(my tensor has >2 dimensions)

**Comment by nikitaved** [comment]
Try the constructor way outlined in my previous message.

**Comment by candyflower2005** [comment]
Out of curiosity - is this value of the scaling factor intentional or is it a bug? https://github.com/pytorch/pytorch/blob/main/torch/sparse/_triton_ops.py#L892
Don't we want to scale it by the square of the number of non-zero elements?
The value was originally set to `sqrt(d_q)` to alleviate the issue of small gradients for large `d_q` but it might vary from row to row.

**Comment by nikitaved** [comment]
It is for now as even the public dense API is not yet fixed. Could you please open a feature request if it is the SDPA you are interested in? It is a trivial addition once approved.

**Comment by candyflower2005** [comment]
Thanks @nikitaved 

Meanwhile, I found something that I would call at least an inconsistency in torch.sparse.
It seems that `torch.sparse_csr` and `torch.sparse_bsr` treat empty columns differently. As a result, I can't create the BSR tensor manually that easily if my data is too sparse. Do you know what the workaround is?

Here's a minimal reproducing example:

```N, p = 1000, 0.0001
t = torch.bernoulli(p * torch.ones((N, N)))

t = t.to_sparse_csr()

bsr = torch.sparse_bsr_tensor(
    crow_indices=t.crow_indices(),
    col_indices=t.col_indices(),
    values=t.values().unsqueeze(-1).unsqueeze(-1),
)
gt_bsr = t.to_dense().to_sparse(layout=torch.sparse_bsr, blocksize=(1,1))

assert torch.equal(gt_bsr.crow_indices(), bsr.crow_indices())
assert torch.equal(gt_bsr.col_indices(), bsr.col_indices())
assert torch.equal(gt_bsr.values(), bsr.values())

print(t.shape)      # torch.Size([1000, 1000])
print(gt_bsr.shape) # torch.Size([1000, 1000])
print(bsr.shape)    # torch.Size([1000, 991])
```

**Comment by candyflower2005** [comment]
Yes, I've tried to do it to convert a CSR tensor to BSR but it didn't work (the error is a few messages above). At the same time, I can't afford doing `t.to_dense().to_sparse(layout=torch.sparse_bsr, blocksize=(1,1))`, because it's too expensive.

**Comment by nikitaved** [comment]
@candyflower2005 , did the constructor path work at least?

**Comment by candyflower2005** [comment]
Yes, thank you for helping me!

**Comment by nikitaved** [comment]
Awesome! Glad to hear :) Will keep it open as it is a legit feature request that could be implemented in no time :)
